### PR TITLE
Grace period for user profile activation

### PR DESCRIPTION
**Pull request by ywangd**
The user profile document is updated on each activate call even when
there is no actual content change because it always updates the
last_synchronized timestamp. This behaviour is intentional to track the
user's last login time (since Kibana calls to the activate API on user
login). Client must explicitly handle retries for version conflicts.
This is generally desirable. However, on each login there are often
multiple web components trying to call this API concurrently. This
results into more frequent version conflicts. Since these updates
occur in a short period of time, updating last_synchronized for each of
them does not really contribute a lot for tracking user login.

This PR introduces a grace period for the update behaviour (30 seconds
non-configurable) so that the update (on activate) is only performed
when either of the following is true:

* There are actual content changes
* Or it has been more than 30 seconds since last update

### CODE REVIEW

File: x-pack/plugin/security/src/main/java/org/elasticsearch/xpack/security/profile/ProfileService.java [hunk 1]
```diff
@@ -687,25 +691,40 @@ private DomainConfig getDomainConfigForSubject(Subject subject) {
...
             buildUpdateRequest(
-                profileDocument.uid(),
-                wrapProfileDocumentWithoutApplicationData(profileDocument),
-                RefreshPolicy.WAIT_UNTIL,
-                versionedDocument.primaryTerm,
-                versionedDocument.seqNo
+                newProfileDocument.uid(),
+                wrapProfileDocumentWithoutApplicationData(newProfileDocument),
+                RefreshPolicy.WAIT_UNTIL
             ),
```
Comment (ywangd) [inline x-pack/plugin/security/src/main/java/org/elasticsearch/xpack/security/profile/ProfileService.java:716]: No behaviour change here. The existing code does not fetch `primaryTerm` and `seqNo` for search result. Hence the net effect is that the update request just runs without a particular expectation for `primaryTerm` and `seqNo`. This is desirable because activate should just rely on the regular update mechanism instead of tying to any specific version. The change here is to make the intention explicit.

**Comment by ywangd** [comment]
@albertzaharovits I just came to realise there are additional complexities in this change. So I changed this PR to draft. You don't need to review it for now. I'll ping you again once it's ready. Sorry for the inconvenience.

File: x-pack/plugin/security/src/main/java/org/elasticsearch/xpack/security/profile/ProfileService.java [hunk 2]
```diff
@@ -687,23 +691,57 @@ private DomainConfig getDomainConfigForSubject(Subject subject) {
...
+                        updateResponse -> new VersionedDocument(
+                            newProfileDocument,
+                            updateResponse.getPrimaryTerm(),
+                            updateResponse.getSeqNo()
+                        ).toProfile(Set.of())
+                    )
+                );
+            }
+        }, listener::onFailure));
+    }
```
Comment (ywangd) [inline x-pack/plugin/security/src/main/java/org/elasticsearch/xpack/security/profile/ProfileService.java:726]: This is the gist of this PR. The code does GET before update so it can reliably tell whether the document has been updated already within past 30 seconds (the critical part is really whether it is updated in the past 1 second because the document-by-search cannot reliably tell that).

**Comment by ywangd** [comment]
The options we have discussed so far (including the current on in this PR) are:
1. For the update operation, catch version conflict exception. If an exception is caught, GET the document and check whether it requires update. If it does *not* require update, return the doc as is. Otherwise, re-throw the exception.
2. GET the document right *before* the update, check whether it requires update. If it requires no update, return the doc as is. Otherweise, go ahead the update.
3. Use scripted update which essentially move the "grace period" logic into the script. 

One thing I'd like to re-emphasize is that the issue is _not_ to avoid "version conflict" because Kibana will retry by default. The issue is to avoid "multiple consecutive version conflict" which can lead Kibana to eventually fail after exhausting all retries. Therefore all of the options can still encounter "version conflict" on first attempt. But their goal is to eliminate "version conflict" on the 2nd attempts so that Kibana is guaranteed to succeed with 3 attempts (1 + 2 retries).

**Review by albertzaharovits — COMMENTED** [review]
In principle I think this is a reasonable approach, but I do have a questions.
First, let me check if I got the problem statement correctly: the problem is that the `_security/profile/_activate` API fails too often with document update conflic errors.
Hence we need to make the internal profile doc update request not fail in order for the `_security/profile/_activate` to be successful.
Also, the doc update must be conditional on the primary_term/seq_no so that the API can work concurrently.

The proosed solution is twofold:
 * use the update timestamp field from the profile doc in order to avoid updating the doc in case it was recently updated
 * use `get` after `search` to retrieve a more recent profile doc which has, in principle, a lower chance for update conflicts

My question is around the second point: if the doc updates use the `wait_until` refresh policy, shouldn't the search already return a recent doc?
In other words, why do we both need the doc `get` and `wait_until` refresh policy for updates?

**Comment by ywangd** [comment]
> if the doc updates use the `wait_until` refresh policy, shouldn't the search already return a recent doc?

The search returns the updated doc only if the searh request comes *after* the `wait_util` is completed. If a concurrent search comes in when another thread is still waiting for its `wait_until`. The search can get back a stale document. The sequence is something like the follows:

0. **T-60.0**: The profile document got updated (`last_synchronized: T-60.0`)
1. **T+0.0**: Thread_1 tries to update the document
2. **T+0.1**: The document is updated by Thread_1 (`last_synchronized: T+0.1`). Thread_1 is still waiting for the operation to complete (`wait_until`) so that the document is *not* visible to search yet.
3. **T+0.2**: Both Thread_2 and Thread_3 come in and try to update the document. They use search to get ahold of the document which is still the old version (`last_synchronized: T-60.0`). 
4. **T+0.3**They _both_ decide to update the document because they see an old timestamp.
5. **T+0.4**  One of Thread_2 and Thread_3 can possibly run into version conflict exception because they try to update the document concurrently. If there are more concurrent threads, the chance of version conflict would be higher (this is what happens with the load tests). This is what turns a "simple" grace period change into a more complicated concurrency issue and where my focus is on.

So the current change proposes that we GET the document in between step 3 and 4. Since GET is realtime, at step 4, both Thread_2 and Thread_3 should find out that the document is in fact recently updated (`last_synchronized: T+0.1`) and skip the update operations, thus avoid version conflict.

Other alternatives are listed [here](https://github.com/elastic/elasticsearch/pull/89566#issuecomment-1228071699). At this point, I am actually thinking option 1 could be better. Because it seems _easier to explain_. Essentially it add a Step 6 to catch the version conflict and GET the document to decide whether it can be suppressed if the document already has the required changes.

**Comment by albertzaharovits** [comment]
> So the current change proposes that we GET the document in between step 3 and 4. Since GET is realtime, at step 4, both Thread_2 and Thread_3 should find out that the document is in fact recently updated (last_synchronized: T+0.1) and skip the update operations, thus avoid version conflict.

I see.
So this solves the version conflicts in the meantime between the successful profile activation and the time when the activation becomes visible to search.

> At this point, I am actually thinking option 1 could be better. Because it seems easier to explain. Essentially it add a Step 6 to catch the version conflict and GET the document to decide whether it can be suppressed if the document already has the required changes.

I agree. I think it would be clearer what's the internal state that the get is designed to serve.
I must admit that, in this draft PR, the search followed by a get looks like a patchwork/best-effort/hope-this-works type of lenient logic.
Option 1) should also be more performant, assuming the time that a profile stays in the "activated but not yet visible for search" state is relatively small.

**Comment by ywangd** [comment]
Thanks Albert! I will change this PR to go with option 1 and ping you again when it is ready.

> Option 1) should also be more performant, assuming the time that a profile stays in the "activated but not yet visible for search" state is relatively small.

I don't think it would be more performant. Though it will have fewer reads (GET),  it will lead to more (unnecessary) writes to the document. In the load test, the difference is 300 vs 14. More specifically, the setup is:
* Check grace period with the search result
* Go ahead with update if the document hasn't been updated for the grace period (based on the search result)
* Update can be scucessful or it can fail with version conflict

The number of successful update is 300 and the number of version conflict is 400. When the code changes to GET doc before update, the number of successful update is only 14 and there is no version conflict. So the difference between Option 1 and 2 should roughly be:
* Option 1(update then catch & get): 300 updates, 400 reads (for error handling)
* Option 2 (get then udpate/skip): 14 updates, 700 reads (before update)

Intuitively, I'd think Option 2 is more performant (especially because every update is a Get+Index). I don't think it is a sufficient argument for Option 2. But I'd like to make the point clear. In theory, Option 3 (scripted update) would be the best of worlds if it is *not* because script can be affected by other settings.

**Comment by albertzaharovits** [comment]
> Intuitively, I'd think Option 2 is more performant (especially because every update is a Get+Index). I don't think it is a sufficient argument for Option 2. But I'd like to make the point clear.

Hmm, Option 1 trades *failed* writes because of `if_seq_no condition` for doc get. We might need to race them, but my intuition is that the failed write is cheaper.

**Comment by albertzaharovits** [comment]
>  I will change this PR to go with option 1 and ping you again when it is ready.

Thank you, I really think it would be an improvement.

**Comment by ywangd** [comment]
@albertzaharovits Thanks again for helping me out on shaping this PR. It's in a much better state now and ready for review!

I ran the load test locally for the current change and I cannot tell any real performance difference. The number of successful logins from Kibana is very close (less than 20 with a total of ~8800). The new approach has much better semantics which is a strong reason for choosing it over the other option. Thanks for the suggestion!

File: x-pack/plugin/security/src/main/java/org/elasticsearch/xpack/security/profile/ProfileService.java [hunk 3]
```diff
@@ -770,25 +772,73 @@ private DomainConfig getDomainConfigForSubject(Subject subject) {
...
+            })
         );
     }
 
+    // If the profile content does not change and it is recently updated within last 30 seconds, do not update it again
+    // to avoid potential excessive version conflicts
+    boolean shouldSkipUpdateForActivate(ProfileDocument currentProfileDocument, ProfileDocument newProfileDocument) {
+        assert newProfileDocument.enabled() : "new profile document must be enabled";
+        if (newProfileDocument.user().equals(currentProfileDocument.user())
+            && currentProfileDocument.enabled()
```
Comment (albertzaharovits) [inline x-pack/plugin/security/src/main/java/org/elasticsearch/xpack/security/profile/ProfileService.java:831]: I know it's probably not possible atm, but activating a profile from `enabled` to `disabled` is liable to skipping which feels wrong to me.

**Review by albertzaharovits — APPROVED** [review]
LGTM
Nice tests, as always.

File: x-pack/plugin/security/src/main/java/org/elasticsearch/xpack/security/profile/ProfileService.java (hunk 3 above)
Comment (ywangd) [reply x-pack/plugin/security/src/main/java/org/elasticsearch/xpack/security/profile/ProfileService.java:831]: I changed it to `newProfileDocument.enabled() == currentProfileDocument.enabled()`. It reads better for semantics.

File: x-pack/plugin/security/src/internalClusterTest/java/org/elasticsearch/xpack/security/profile/ProfileDomainIntegTests.java [hunk 4]
```diff
@@ -341,6 +345,60 @@ public void testConcurrentCreationOfNewProfiles() throws InterruptedException {
...
+
+        if (readyLatch.await(20, TimeUnit.SECONDS)) {
+            startLatch.countDown();
+            for (Thread thread : threads) {
+                thread.join();
+            }
+        }
+
+        assertThat(updatedProfiles, not(emptyIterable()));
+        final Profile updatedProfile = updatedProfiles.stream().max(Comparator.comparingLong(Profile::lastSynchronized)).orElseThrow();
```
[the remark this answers is not in the document]
Comment (ywangd) [reply x-pack/plugin/security/src/internalClusterTest/java/org/elasticsearch/xpack/security/profile/ProfileDomainIntegTests.java:395]: I should have commented on this block of code because it is not obvious. Due to the concurrency nature, we cannot assert that all activated profiles are equal (because of `last_synchronized`). The version conflict error is not guarantee to happen. It is entirely possible that two concurrent updates both go through without any error which means the profile document gets updated twice. The bottom line is that we know at least one of the update will go through without the need for error handling. The other may go through normally or they may fail with version conflict which is handled and suppressed. So what we can assert is that all updates are successful (one way or another) rather than the actual updated content. 

I added a comment in the code to explain this.

File: x-pack/plugin/security/qa/profile/src/javaRestTest/java/org/elasticsearch/xpack/security/profile/ProfileIT.java [hunk 5]
```diff
@@ -352,6 +358,56 @@ public void testXpackUsageOutput() throws IOException {
...
+        try {
+            // Initial activate with node0
+            client.setNodes(List.of(node0));
+            final Map<String, Object> responseMap0 = responseAsMap(client.performRequest(activateProfileRequest));
+
+            final Instant start = Instant.now();
+            // Activate again with the same host (node0) should fall within the grace period and skip actual update
+            final Map<String, Object> responseMap1 = responseAsMap(client.performRequest(activateProfileRequest));
+            assertThat(responseMap1.get("_doc"), equalTo(responseMap0.get("_doc")));
+            assertThat(start.plus(30, ChronoUnit.SECONDS).isAfter(Instant.now()), is(true));
```
[the remark this answers is not in the document]
Comment (ywangd) [reply x-pack/plugin/security/qa/profile/src/javaRestTest/java/org/elasticsearch/xpack/security/profile/ProfileIT.java:394]: Yeah `assumeTrue` is better. I replace it here and in other two places. Thanks!

File: x-pack/plugin/security/src/main/java/org/elasticsearch/xpack/security/profile/ProfileService.java [hunk 6]
```diff
@@ -770,25 +772,73 @@ private DomainConfig getDomainConfigForSubject(Subject subject) {
...
+                }
+            })
         );
     }
 
+    // If the profile content does not change and it is recently updated within last 30 seconds, do not update it again
+    // to avoid potential excessive version conflicts
+    boolean shouldSkipUpdateForActivate(ProfileDocument currentProfileDocument, ProfileDocument newProfileDocument) {
+        assert newProfileDocument.enabled() : "new profile document must be enabled";
+        if (newProfileDocument.user().equals(currentProfileDocument.user())
```
[the remark this answers is not in the document]
Comment (ywangd) [reply x-pack/plugin/security/src/main/java/org/elasticsearch/xpack/security/profile/ProfileService.java:830]: We can definitely adjust this in a future PR if necessary. I chose the current behaviour for safety over efficiency. Also like you said, the code looks quite ugly when we cannot simply rely on `user.equals`. The logic is simple enough though. What we have in this PR has done the harder part. So it should be fairly easy to adjust for the equality check. I'll circle this back to the kibana team and maybe raise a follow-up PR. Thanks!

File: x-pack/plugin/security/src/internalClusterTest/java/org/elasticsearch/xpack/security/profile/ProfileDomainIntegTests.java (hunk 4 above)
[the remark this answers is not in the document]
Comment (albertzaharovits) [reply x-pack/plugin/security/src/internalClusterTest/java/org/elasticsearch/xpack/security/profile/ProfileDomainIntegTests.java:399]: Thank you, I understand it now.
